### DISCUSSION TITLE
USB N2K gateway

### DISCUSSION THREAD

**Discussion by sass00** [hatlabs/discussions#63] [General]
Hey guys,

I allready posted about my issue at OpenMarine forum but it seems this might be the best adress. I have my N2k gateway paired to RPi4B with OpenPlotter3. 

My backbone is done with both terminating resistors and my E80 Raymarine is the only device on the bus. The issue is i can see the output of the SK-N2k plugin with PGN's and everything but after i open signalk:stream canboatjs there is no output. The funny thing is that my TX counter is actually moving on the interface itself (on the OLED) but my raymarine is not detecting any device on the bus or seems to not be getting anything sent through the bus. 

I have tried setting canboat/canboatjs, different baud rates, ... tried disabling enabling, changing USB ports, cables, ... 

I have the gateway for some time now but spent some time installing it on my boat now and configured SK on the boat also. Any advices what else could i try? 

I have the engine hat also connected and i can see also the engine parameters PGN's being outputted from the SK to N2k plugin, pressure, sog&cog, position, ... but nothing makes it to my plotter actually. I tried dismantling the enclosure today to check for any wrong wiring on the N2k connector i ended up soldering the N2k header (from the ESP32) directly to the N2k cable and tried again - same result. 

Any advices or ideas what to check? If i open CAN-bus app on the OP3 and try unblocking PGN's it also says that i should reset the device or change the baudrate since no PGN's are enabled.

**Comment by nousian** [comment]
How about try to generate some synthetic N2K messages from the device (gw) itself?

I use the following function to feed wind (coming from elsewhere) into N2K bus:

```
void updateWind(const double windangleDeg, const double windspeed) { 
  if (bus_status == 0) {
    const double windangle = DegToRad(windangleDeg);

    tN2kMsg message;
    SetN2kPGN130306(message, UC255, windspeed, windangle, N2kWind_Apparent);
    HandleStreamActisenseMsg(message);
  }  
}
```

All my instruments pick that just fine, wind repeater, plotter, autopilot, nav computer etc.

Try feeding some random stuff like that into bus and see if your plotter pics those up. Might be plotter settings, my Furuno TZ needs to manually select source for each data type, by default it won't "see" N2K data.

**Comment by sass00** [comment]
Im not really good with writing code. I'm looking for a way to feed it a dummy position to see if it recieves it. But as im looking through N2kMessages.h i'm not really sure how to do it.

**Comment by nousian** [comment]
> Im not really good with writing code. I'm looking for a way to feed it a dummy position to see if it recieves it. But as im looking through N2kMessages.h i'm not really sure how to do it.

I would suggest something more simple than position to test out dummy messages, like wind as in my example or depth or speed etc.

See position message looks pretty complicated, and I guess you have to get most right in order for the plotter to accept it:

```
void SetN2kPGN129029(tN2kMsg &N2kMsg, unsigned char SID, uint16_t DaysSince1970, double SecondsSinceMidnight,
                     double Latitude, double Longitude, double Altitude,
                     tN2kGNSStype GNSStype, tN2kGNSSmethod GNSSmethod,
                     unsigned char nSatellites, double HDOP, double PDOP=0, double GeoidalSeparation=0,
                     unsigned char nReferenceStations=0, tN2kGNSStype ReferenceStationType=N2kGNSSt_GPS, uint16_t ReferenceSationID=0,
                     double AgeOfCorrection=0
                     );
```
